feat: new InstructionList for instructions / requirements

sbrandwoo opened the pull request:
![Selection_051](https://user-images.githubusercontent.com/614392/70144356-a6a52380-1695-11ea-8b19-93e12454fb7e.png)

## ❓ Context <!-- why this change is made --> 

When opening bank details and ordering cards there is now a new requirements list, much like the existing instructions list on mobile.

## 🚀 Changes <!-- what this PR does -->

Adds `InstructionList`.

## 💬 Considerations <!-- additional info for reviewing, discussion topics -->

This PR should probably also remove `InstructionsList`, which could be better named as `DosAndDontsList`. I'll await feedback on the new component before doing this.

The TypeScript definitions do not make it into the output bundle :(
I'm not aware how to do this in the current build flow.

## ✅ Checklist

- [ ] All changes are covered by tests
- [x] The changes are covered in docs
- [x] The commits follow conventional commits standards

oliverviljamaa commented on packages/components/CHANGELOG.md:
```diff
@@ -752,7 +752,7 @@ Fix failing test and make code more clear
 
 # v4.15.0
-## Add InstructionList component
+## Add InstructionsList component
```
Oops.

oliverviljamaa commented on packages/components/src/instructionList/InstructionList.less:
```diff
@@ -0,0 +1,34 @@
+@import 'bootstrap/less/variables.less';
```
@nabaroa Should we import from `neptune-css` here already? If yes, do we have an example?

dozortsev replied:
I guess these styles also should be in neptune/css package

dozortsev replied:
re. variables

yes, I think you can import them but not all variables are there yet https://github.com/transferwise/neptune/blob/master/packages/css/src/less/variables/_variables.less

I would try import it and check current neptune `_variables.less` satisfy needs

dozortsev replied:
here is the example of importing color variables from neptune https://github.com/transferwise/neptune/blob/91fdd6a3b0124c6127e33e03df6f3b52e562a716/packages/components/src/tabs/Tabs.less#L2

sbrandwoo replied:
Really? Why?

oliverviljamaa commented on packages/components/src/instructionList/index.d.ts:
```diff
@@ -0,0 +1,17 @@
+import React, { ReactNode } from 'react';
+
+export interface Instruction {
```
👏

oliverviljamaa commented on packages/components/src/instructionList/index.d.ts:
```diff
@@ -0,0 +1,17 @@
+  instructions: Instruction[];
+}
+
+export const InstructionList: React.FunctionComponent<InstructionListProps>;
```
😍

gusso replied:
omg yes

sbrandwoo replied:
I discovered an unfortunate side-effect. If I copy this file into `/node_modules/@transferwise/components` then TypeScript is like "OMG I know `InstructionList`!"  - which is great. But then it thinks that all other components don't exist because there's no types of them. It'd usually just apply `Button` as `any`, but with this in place it'll be an error.

I don't know if there's a way to tell TypeScript that this is just some of the types, and to import anything else as `any`?

oliverviljamaa commented on packages/components/src/instructionList/index.js:
```diff
@@ -0,0 +1,3 @@
+import InstructionList from './InstructionList';
```
This is often used across the company, as it allows us to omit the intermediate variable:
```suggestion
export { default } from './InstructionList';
```

oliverviljamaa commented on packages/components/src/instructionList/index.d.ts:
```diff
@@ -0,0 +1,17 @@
+  icon: ReactNode;
+  text: ReactNode;
+  help?: {
+    title?: ReactNode | null;
```
`null` is a valid `ReactNode`.
```suggestion
    title?: ReactNode | null;
```

sbrandwoo replied:
Hmm, I don't want that. `icon` and `text` shouldn't allow null :thinking:

oliverviljamaa commented on packages/components/src/instructionList/InstructionList.less:
```diff
@@ -0,0 +1,34 @@
+
+    &__body {
+      margin-left: 40px;
+      font-size: 14px;
```
Can we use variables for these as well?

sbrandwoo replied:
What would I use?

oliverviljamaa replied:
`@font-size-small` for `14px` font and `@spacer * x` for spacing (`@spacer` is `8px`).

sbrandwoo replied:
thanks!

oliverviljamaa commented on packages/components/src/instructionList/InstructionList.js:
```diff
@@ -0,0 +1,51 @@
+  return <div className="instruction-list">{items}</div>;
+};
+
+InstructionList.propTypes = {
```
Would be cool if we could somehow ensure `propTypes` and the types will always be aligned 🤔

oliverviljamaa replied:
The simplest way of doing it would probably be to make this a TypeScript file and use `FC` with defined prop types, but adding TypeScript to this repo is a choice we're probably not ready to make yet?

oliverviljamaa commented on packages/components/src/instructionList/InstructionList.js:
```diff
@@ -0,0 +1,51 @@
+    );
+  });
+
+  return <div className="instruction-list">{items}</div>;
```
Do you think the `items` variable is necessary?

andreapiras replied:
div can wrap directly the map

sbrandwoo replied:
This is a very mild case, but I'm not a huge fan of nesting absolutely everything inside a single statement. So if there's a `.map` or other complex structure, I'll split it out for readability.

oliverviljamaa commented on packages/components/src/instructionList/InstructionList.js:
```diff
@@ -0,0 +1,51 @@
+        <div className="instruction__main">
+          <div className="instruction__icon m-r-2">{icon}</div>
+          <div className="instruction__text">
+            <div>
```
Do we need that `div`?

gusso commented on packages/components/src/instructionList/InstructionList.docs.js:
```diff
@@ -0,0 +1,93 @@
+            <br />
+            <a href="#foo">Edit address</a>
+          </p>
+        </>
```
This example doesn't need a fragment, right? Are you including it to encourage every body property to have one?

sbrandwoo replied:
Just bad copy-pasting

andreapiras commented on packages/components/src/index.js:
```diff
@@ -16,6 +16,7 @@ export { default as DateLookup } from './dateLookup';
 export { default as PhoneNumberInput } from './phoneNumberInput';
 export { default as DateInput } from './dateInput';
 export { default as InstructionsList } from './instructionsList';
+export { default as InstructionList } from './instructionList';
```
How come we are not modifying the existing InstructionsList. I find this InstructionsList vs InstructionList etxremely confusing

andreapiras commented on packages/components/src/instructionList/InstructionList.js:
```diff
@@ -0,0 +1,51 @@
+          </div>
+          {help && (
+            <div className="instruction__help m-l-2">
+              <Popover preferredPlacement="top" title={help.title} content={help.content}>
```
maybe deconstruct help and content?

andreapiras commented on packages/components/src/instructionList/InstructionList.js:
```diff
@@ -0,0 +1,51 @@
+          {help && (
+            <div className="instruction__help m-l-2">
+              <Popover preferredPlacement="top" title={help.title} content={help.content}>
+                <HelpCircleIcon size="sm" />
```
Do we always need this? can we add a propr to control whether the helpcircle is there or not?

dozortsev commented on packages/components/src/instructionList/InstructionList.js:
```diff
@@ -0,0 +1,49 @@
+
+import './InstructionList.css';
+
+const InstructionList = ({ instructions }) => (
```
according to this [rfc](https://github.com/transferwise/neptune/blob/master/packages/docs/pages/rfc/0001-always-pass-classname.md) please expose `className` prop 🙂

dozortsev replied:
in most outer DOM element

sbrandwoo commented:
Outstanding questions for the component interface:
- Should the text be by default in `<strong>` elements, or do we let the parent decide?
- Should the left icon be forced to be navy, or do we let the parent decide?
Both are difficult to overwrite if you don't want them. If we want to support red and green icons, then we'll need to do _something_...

sbrandwoo commented:
Source design, for reference:

![image](https://user-images.githubusercontent.com/614392/70332830-4996a280-183a-11ea-9aa2-4256a3b340ef.png)

sbrandwoo commented:
@dozortsev I've addressed the popover bug in https://github.com/transferwise/neptune/pull/122 and changed it to open bottom-left so that it's more likely to fit on the screen.

oliverviljamaa commented on packages/components/src/instructionList/InstructionList.js:
```diff
@@ -0,0 +1,53 @@
+
+import './InstructionList.css';
+
+const InstructionList = ({ className, instructions }) => (
```
Is it always going to be an _instruction_ list? It's icons, texts, and statuses, so it could be used for more than instructions?

gusso commented:
Replaced with #319
